### ISSUE TITLE
cf auth fails with dial tcp i/o timeout

### ISSUE THREAD

**Issue by t-seyama** [cloudfoundry/cli#1286]
## Command  

```cf auth```

## What occurred
Dial tcp i/o timeout rarely occurs.
```
cf auth admin mypasswaord
API endpoint: https://api.mycf.com
Authenticating...
Post http://uaa.mycf.com/oauth/token: dial tcp: i/o timeout
FAILED
```

## CLI Version  

cf version 6.32.0+0191c33d9.2017-09-26

## CC API Endpoint Version  

2.97.0

## Platform & Shell Details  

Ubuntu 14.04.1 LTS

## Any other relevant information  

As a result of packet capture, the CLI seemed to be waiting for the response of the AAAA record and timed out when the response got lost.

success
```
664281	2017-12-13 09:36:10.565450	10.0.255.4	8.8.8.8	DNS	98	Standard query 0x5956 AAAA uaa.mycf.com
664282	2017-12-13 09:36:10.565541	10.0.255.4	8.8.8.8	DNS	98	Standard query 0xeb87 A uaa.mycf.com
664283	2017-12-13 09:36:10.572708	8.8.8.8	10.0.255.4	DNS	180	Standard query response 0x5956 AAAA uaa.mycf.com SOA ns-1255.awsdns-28.org
664284	2017-12-13 09:36:10.572720	8.8.8.8	10.0.255.4	DNS	114	Standard query response 0xeb87 A uaa.mycf.com A 10.0.0.100
```

failure
```
664306	2017-12-13 09:36:10.587869	10.0.255.4	8.8.8.8	DNS	98	Standard query 0xb1d4 AAAA uaa.mycf.com
664307	2017-12-13 09:36:10.587956	10.0.255.4	8.8.8.8	DNS	98	Standard query 0xcd19 A uaa.mycf.com
664309	2017-12-13 09:36:10.596365	8.8.8.8	10.0.255.4	DNS	114	Standard query response 0xcd19 A uaa.mycf.com A 10.0.0.100
```

It looks like #1080.

**Comment by dkoper** [comment]
Hi @t-seyama 

cf CLI 6.33.0 is built on a later Golang version (1.9.2 instead of 1.9.0 for the 6.32.0 version you are using).
Does the issue occur on this latest CLI version as well?

Cheers,
Dies Koper
CF CLI PM

**Comment by t-seyama** [comment]
Hi @dkoper 

I updated to cf CLI 6.33.0, but the issue also occurred.

**Comment by dkoper** [comment]
Thanks for confirming. I'll ask my team to take a look.

**Comment by nickwei84** [comment]
Hi @t-seyama, we're trying to reproduce this issue in our own environment. Can you please provide the following:
- are you doing any special setup to produce this timeout error
- what tool are you using to capture the DNS logs
- are you using the standard UAA server setup or a custom one
- it looks like your UAA server is in the same private network as your CC server?

and anything else you can think of to help us reproduce this timeout error. Thanks.

**Comment by t-seyama** [comment]
Hi @nickwei84

> are you doing any special setup to produce this timeout error

I continued to execute cf auth command from multiple clients.

> what tool are you using to capture the DNS logs

I used tshark and caputured packets on the client side.

> are you using the standard UAA server setup or a custom one

Yes, I set up UAA using cf-release. 

> it looks like your UAA server is in the same private network as your CC server?

Yes, they are in the same private network.


Thanks.

**Comment by dkoper** [comment]
Hi @t-seyama 

The team spent a few hours investigating but it seems we've reached our limits of what we can do to figure this out. In our efforts to reproduce, it seems to occur only once in 180 times. You also mentioned it occurs only rarely.

I think it's better that we give up here and go back to coding features.

If you want to investigate further, note that our Linux CLI binary uses the go resolver, so you could try rebuilding it with the cgo resolver, or you could try analysing the debugging output when setting Golang's `GODEBUG=netdns=1` environment variable.

If you find anything that points to a bug in our code, please let us know and we'll pick it up again.
Else, we'll assume this is a Golang issue and hope a future Golang release will resolve it.

**Comment by t-seyama** [comment]
Hi @dkoper 

Thank you for investigation.
I will continue to investigate. I inform you if there is any new information.

Thanks.

**Comment by dkoper** [comment]
@t-seyama Thanks!

**Comment by XenoPhex** [comment]
Hey @t-seyama thanks for creating this issue. We are closing this issue due to inactivity but please feel free to reopen or create a new issue if you have feedback.

For anyone stumbling on this after we close it. The CF CLI Team has found that these errors tend to come up due to DNS misconfigurations or DNS Timeouts. All of which boil down to "DNS is not configured as your platform expects it". If you can fix the DNS configuration, this will no longer be an issue, if you can't/are unable to figure out what the problem is. Try setting the `CF_DIAL_TIMEOUT` environment variable to a value higher than the default value of `5` (in seconds).
